gpt_index/docstore.py

```diff
@@ -8,6 +8,7 @@
 
 from gpt_index.data_structs.data_structs_v2 import V2IndexStruct as IndexStruct
 from gpt_index.data_structs.node_v2 import Node
+from gpt_index.indices.registry import INDEX_STRUCT_TYPE_TO_INDEX_STRUCT_CLASS
```

Comment:
are there circular import issues here? 

docstore imports from indices.registry
indices.registry imports from every index class
every index class inherits from base index class
base index class imports docstore

Reply:
a few ways around this: 
(1) make mapping from index type to index struct class / query map directly (don't import outer index class)
(2) abstract out a "BaseDocstore" interface and pass that into BaseGPTIndex instead of the full Docstore

i mildly prefer the former since it's a bit weird conceptually to have docstore depend on the outer index class but curious for your thoughts

Reply:
yep, good catch, only saw this while testing the next PR.

Reply:
Will revise to 1)

Reply:
Actually, revising to have mapping from index type to query map directly means I have to create a central dict of query map of all indices again, is this desirable?

Reply:
Another hacky option is to do lazy import of `INDEX_STRUT_TYPE_TO_QUERY_MAP` and `INDEX_STRUCT_TYPE_TO_INDEX_STRUCT_CLASS`, so that they are not imported at module import time.

Reply:
> Actually, revising to have mapping from index type to query map directly means I have to create a central dict of query map of all indices again, is this desirable?

Hm yeah that's true.

On a general note why do we `need` to have this line `type_to_struct = type_to_struct or INDEX_STRUCT_TYPE_TO_INDEX_STRUCT_CLASS`? the only module that needs the hardcoded dict is our composable graph. If the user wanted to initialize the docstore themselves (which we don't really support yet), they could pass in the dict themselves. within a single index class, we don't need the hard-coded map. 

i guess that was the point of having an "index registry" object in the first place. the main change here is removing the dataclass and replacing with a hardcoded dict

Reply:
I think the motivation for removing `IndexRegistry` is that I don't think it should be customized by the user.

It's simplify a mapping between an identifier `IndexStructType` to the actual`IndexStruct` class. I don't get why it should be customized.

Reply:
if the user adds their own custom index maybe?